### PR TITLE
go node: unite compute, merge, and transaction scheduler roles

### LINKED ISSUE #2107: Remove granular role registration
_Re: Core Platform Meeting (2019/09/17)_

The registry currently allows granular registration of roles within the compute committee but this should be changed so that the per-runtime compute execution roles (txnscheduler/compute/merge nodes) are now a single "compute" role from the viewpoint of the registry.

The scheduler should still schedule individual roles as-is and would pick from all valid nodes registered with the new "compute" role.

### PR DESCRIPTION
**Pull request by pro-wh**
We're removing the separation among registering nodes for the compute, merge, and transaction scheduler roles. You now have to register for and enable all or none of these roles, under a new, broadened, and confusing--you're welcome--term "compute."

non-changelog data:

- the `.enabled` flag is moved to its own package, `computeenable`. this is because there are dependencies leading toward the merge worker package, but that would be a weird place to put the compute enable flag.
- a backwards incompatible change is made to the on-chain representation of registered nodes, and no migration is provided. you're welcome.

closes #2107

### CODE REVIEW

**Comment by pro-wh** [comment]
aw this is going to break the byzantine tests isn't it

File: go/consensus/tendermint/apps/scheduler/scheduler.go [hunk 1]
```diff
@@ -393,7 +393,7 @@ func (app *schedulerApplication) isSuitableTransactionScheduler(n *node.Node, rt
 }
 
 func (app *schedulerApplication) isSuitableMergeWorker(n *node.Node, rt *registry.Runtime, ts time.Time) bool {
-	return n.HasRoles(node.RoleMergeWorker)
+	return n.HasRoles(node.RoleComputeWorker)
```
Comment (kostko) [inline go/consensus/tendermint/apps/scheduler/scheduler.go]: This (and storage) needs to be made per-runtime (as mentioned in https://github.com/oasislabs/oasis-core/issues/2107#issuecomment-570572948), similar to txnscheduler/compute.

File: .changelog/2107.breaking.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+go node: Unite compute, merge, and transaction scheduler roles.
+
+We're removing the separation among registering nodes for the compute, merge, and transaction scheduler roles.
+You now have to register for and enable all or none of these roles, under a new, broadened, and confusing--you're
+welcome--term "compute."
```
Comment (kostko) [inline .changelog/2107.breaking.md]: Yeah this is quite confusing, maybe we should make it so:
* The overall term for all three is "compute".
* The old "compute" is renamed to "executor" (or something, no strong preference here).

Thoughts?

File: go/worker/computeenable/init.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package computeenable
```
Comment (kostko) [inline go/worker/computeenable/init.go]: Ugh, this seems ugly. In light of my previous comment around naming we could also do something like this:
* Rename existing `compute` to `executor` (or some similar name as mentioned above).
* Restructure workers to have `go/worker/compute/{txnscheduler,executor,merge}`.
* Have this "computeenable" thing in `go/worker/compute` potentially with some common code that initializes all three workers.

File: go/consensus/tendermint/apps/scheduler/scheduler.go (hunk 1 above)
Comment (pro-wh) [reply go/consensus/tendermint/apps/scheduler/scheduler.go]: ok I added the runtime snippet to these

File: .changelog/2107.breaking.md (hunk 2 above)
Comment (pro-wh) [reply .changelog/2107.breaking.md]: ya ok

Comment (pro-wh) [reply .changelog/2107.breaking.md]: #2525

File: go/worker/computeenable/init.go (hunk 3 above)
Comment (pro-wh) [reply go/worker/computeenable/init.go]: #2525 for the first 2 bullet points

**Comment by pro-wh** [comment]
remaining weirdness:
- registry node runtimes are shared across multiple roles
- some tests might not be up to date on per-runtime merge and storage change

File: go/worker/computeenable/init.go (hunk 3 above)
Comment (pro-wh) [reply go/worker/computeenable/init.go]: moved computeenable to compute. initialization stuff is mostly in executor, with that being the one that needs a registration callback (to wait for TEE capability)

File: go/consensus/tendermint/apps/beacon/beacon.go [hunk 4]
```diff
@@ -122,6 +122,7 @@ func (app *beaconApplication) onBeaconEpochChange(ctx *abci.Context, epoch epoch
 	case true:
 		// UNSAFE/DEBUG - Deterministic beacon.
 		entropyCtx = debugEntropyCtx
+		entropy = []byte("luck=1")
```
Comment (kostko) [inline go/consensus/tendermint/apps/beacon/beacon.go]: Why is this required? Isn't it the same if you just tweak the identity seeds (with different values ofc)? If it is not, add a comment why this is required.

Comment (pro-wh) [reply go/consensus/tendermint/apps/beacon/beacon.go]: I'll add a comment

**Comment by pro-wh** [comment]
alright, I'd better take inventory of what went down in coverage

1. StateProcessingBatch.cancel
2. Namespace.UnmarshalBinary ErrMalformedNamespace
3. Namespace.UnmarshalText error from base64 DecodeString
4. Namespace.UnmarshalHex error from hex.DecodeString
5. Namespace.Equal `cmp == nil`
6. cmd debug storage check-roots Args error from nrFn
7. ibid, error from ValidateRuntimeIDStr
8. cmd debug storage force-finalize Args
9. cmd debug storage ValidateRuntimeIDStr error from hex.DecodeString
10. cmd debug storage checkDiff error from storageClient.GetDiff
11. ibid, error from write log iterator Next
12. ibid, error from write log iterator Value
13. cmd debug storage doCheckRoots error from storageClient.New
14. ibid, error from client.GetBlock
15. ibid, error from storageWorkerClient.GetLastSyncedRound
16. ibid, retry when storage worker is not synced
17. ibid, error from client.GetBlock
18. cmd debug storage doForceFinalize `failed` branch

I'm stopping here. Observations:
- covered -> not covered transitions are almost all error handling branches, which I suspect are ones that we don't explicitly exercise in tests
- item 18 https://codecov.io/gh/oasislabs/oasis-core/pull/2514/changes#L282 ought to be impossible, because you would have to go through line 247 (never covered) before you could get to line 282 (previously covered, now not covered)
- what is force-finalize? do we exercise it?

**Comment by pro-wh** [comment]
now the runtime-dynamic test doesn't work, because you wouldn't be able to register per-runtime storage nodes until the runtimes are registered

**Comment by pro-wh** [comment]
alright, adjusted as the comment there described
